### ISSUE TITLE
Accessibility: CSS :before and :after pseudo-elements insert non-decorative content

### ISSUE THREAD

**Issue by vikrampm1** [woothemes/woocommerce#51795]
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The CSS :before and :after pseudo-elements are used to add non-decorative content before and after an element's content in the document tree.

CSS :before and :after pseudo-elements specify the location of content before and after an element's document tree content. The content property, in conjunction with these pseudo-elements, specifies what is inserted.

Users who need to customize style information to view content according to their needs may not be able to access information inserted via CSS.

It is an accessibility failure to use these properties to insert non-decorative content such as the visual indicator of an input's checked status.

![T001_SR_CheckboxDefinedThroughCSS](https://github.com/user-attachments/assets/d47c44e8-84b5-4d99-b7b0-15310836b252)

![T001_SR_RadioBtnDefinedThroughCSS](https://github.com/user-attachments/assets/e85194af-9245-437c-9051-5ff32dc0b55c)

![T001_SR_CheckboxDefinedThroughCSS2](https://github.com/user-attachments/assets/ab131de5-1db8-4536-bb02-65ed234bc3f4)

### Component or Page: 

**Page: Checkout Process (Block-Based)**

### Name affected element:

1. Radio buttons, such as:
- Direct bank transfer
- Check payments
- Cash on delivery
2. Add a note to your order checkbox

### Relevant code snippets:

```
<input id=""radio-control-wc-payment-method-options-bacs"" class=""wc-block-components-radio-control__input"" type=""radio"" name=""radio-control-wc-payment-method-options"" aria-describedby=""radio-control-wc-payment-method-options-bacs__label"" aria-disabled=""false"" value=""bacs"" checked="""">

CSS:
.wc-block-components-radio-control .wc-block-components-radio-control__input:checked:before {
    background: #000;
    border-radius: 50%;
    content: """";
    display: block;
    height: .75em;
    left: 50%;
    margin: 0;
    min-height: 12px;
    min-width: 12px;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    width: .75em;
}

```
### Recommended fix:

Ensure that CSS :before and :after pseudo-elements are not employed for inserting non-decorative content.

Restyle inputs to have an alternate way of displaying checked status that doesn't solely rely on CSS :before and :after attributes.

**Comment by mikejolley** [comment]
It was my understanding that the `input` was still part of the DOM and this is in sync with the visual styling. Do the screenreaders not reference that element to read if the input is checked/unchecked?

**Comment by jeffpaul** [comment]
@amberhinds please see question above to help clarify this issue

**Comment by amberhinds** [comment]
I'm double-checking this one and will follow up this afternoon. I need to confirm how it functions in a few different screen readers.

**Comment by amberhinds** [comment]
Sorry for the delay. Just noting this is still on my radar.

**Comment by jeffpaul** [comment]
@amberhinds reminder ping (especially before the GitHub Bot closes this on us)

**Comment by amberhinds** [comment]
I have tested this in Safari and Chrome with VoiceOver, and Firefox with NVDA and can confirm that it sounds correct in those environments. There is a visible focus and appropriate keyboard interactions on all instances of these inputs. I have also disabled CSS on the checkout page and confirmed that the normal input is returned with a visual indicator of what has been selected and the selected state is announced properly without the presence of the :before styles. 

Tentatively, I believe that this is a non-issue and should not have been flagged. However, I have asked @digitala11y to review this item as he originally flagged it. I want to make sure this is not an issue with JAWS, which he has and I do not.

**Comment by amberhinds** [comment]
I'll have an answer on this today.

The short of this is that we have an internal discussion about the fact that the state of the input is not visible if the :before styles don't load, as shown in this video.

https://github.com/user-attachments/assets/fc525c01-0b73-4e14-a546-0039a49822dd

However, I'm not certain that this represents a real-world scenario where _only_ those styles don't load. If all the styles don't load, it works normally. Thus, I expect that it might not be an issue at all and I'm doing some digging this morning to see if I can identify a scenario where the :before styles (but no others) are blocked.

@mikejolley or @jeffpaul can you think of a scenario when a user would encounter what's shown in the video (outside of manually disabling the styles for that one selector in the inspector as shown)?

**Comment by jeffpaul** [comment]
I can't think of a scenario where that happens besides someone forcing it to happen that way (or maaaaybe perhaps a network/local device hiccup and some of those styles didn't load) but for the most part, no, I cannot perceive this sort of scenario.

**Comment by amberhinds** [comment]
Yes, that's what I was thinking. We should close this issue and consider it a false positive.

**Comment by NicktheGeek** [comment]
Hey all, I just want to add some additional context on this. 

TLDR; I agree with @amberhinds's conclusions. Just spelling it out in case someone comes here later and wonders about it.

It looks like this is related to [WCAG Success Criterion 1.3.1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html):
> 1.3.1 Info and Relationships: Information, [structure](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#structuredef), and [relationships](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relationshipsdef) conveyed through [presentation](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#presentationdef) can be [programmatically determined](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#programmaticallydetermineddef) or are available in text. (Level A)

More specifically, [technique F87](https://www.w3.org/TR/WCAG20-TECHS/F87.html):
> Failure of Success Criterion 1.3.1 due to inserting non-decorative content by using :before and :after pseudo-elements and the ‘content’ property in CSS
>
> **Description**
> The CSS :before and :after pseudo-elements specify the location of content before and after an element’s document tree content. The content property, in conjunction with these pseudo-elements, specifies what is inserted. For users who need to customize or turn off style information in order to view content according to their needs, assistive technologies may not be able to access the information that is inserted using CSS. Therefore, it is a failure to use these properties to insert non-decorative content.

This is specifically about having actual content that conveys information in the content css property. In this case there is no actual content. It’s an empty string. There is “information” conveyed through the styling but as noted in the description, turning off styles would mean the default radio input would show. Additionally, since the “checked” state is on the an HTML input element, this state is conveyed to screen reader users.

So this is not an accessibility failure. This is a case of enhanced styling following appropriate methods:
- Semantic markup is used (ie <input>  and <label>)
- Default functionality is NOT overridden
- Styles are used to enhance the visual experience
- Since semantic markup is used, screen reader users have full access to the visual presentation
- Since semantic markup is used, should styles be disabled by the user, the default functionality would continue to work as expected

I hope that if someone finds this in the future, the added description here will help clarify why this is not an accessibility failure.

**Comment by amberhinds** [comment]
Thanks @NicktheGeek!
